aten/src/ATen/core/Backend.h

```diff
@@ -0,0 +1,168 @@
+  switch (b) {
+    case Backend::CPU:
+      return Backend::CPU;
+    case Backend::CUDA:
+      return Backend::CUDA;
+    case Backend::SparseCPU:
+      return Backend::CPU;
+    case Backend::SparseCUDA:
+      return Backend::CUDA;
+    default:
+      throw std::runtime_error("Unknown backend");
+  }
+}
+
+static inline Backend tensorTypeIdToBackend(TensorTypeId t) {
+  if (t == CPUTensorId()) {
+    return Backend::CPU;
+  } else if (t == CUDATensorId()) {
+    return Backend::CUDA;
+  } else if (t == SparseCPUTensorId()) {
+    return Backend::SparseCPU;
+  } else if (t == SparseCUDATensorId()) {
+    return Backend::SparseCUDA;
+  } else if (t == UndefinedTensorId()) {
+    return Backend::Undefined;
+  } else {
+    AT_ERROR("Unrecognized tensor type ID: ", t);
+  }
+}
+
+static inline TensorTypeId backendToTensorTypeId(Backend b) {
+  switch (b) {
+    case Backend::CPU:
+      return CPUTensorId();
+    case Backend::CUDA:
+      return CUDATensorId();
+    case Backend::SparseCPU:
+      return SparseCPUTensorId();
+    case Backend::SparseCUDA:
+      return SparseCUDATensorId();
+    case Backend::Undefined:
+      return UndefinedTensorId();
+    default:
+      throw std::runtime_error("Unknown backend");
+  }
+}
+
+static inline DeviceType backendToDeviceType(Backend b) {
+  switch (b) {
+    case Backend::CPU:
+      return DeviceType::CPU;
+    case Backend::CUDA:
+      return DeviceType::CUDA;
+    case Backend::SparseCPU:
+      return DeviceType::CPU;
+    case Backend::SparseCUDA:
+      return DeviceType::CUDA;
+    case Backend::Undefined:
+      AT_ERROR("Undefined backend is not a valid device type");
+    default:
+      AT_ERROR("Unknown backend");
+  }
+}
+
+static inline Backend deviceTypeToBackend(DeviceType d) {
+  switch (d) {
+    case DeviceType::CPU:
+      return Backend::CPU;
+    case DeviceType::CUDA:
+      return Backend::CUDA;
+    default:
+      AT_ERROR("Unknown device type ", d);
+  }
+}
+
+static inline Backend backendToCPU(Backend b) {
+  switch (b) {
+    case Backend::CPU:
+      return Backend::CPU;
+    case Backend::CUDA:
+      return Backend::CPU;
+    case Backend::SparseCPU:
+      return Backend::SparseCPU;
+    case Backend::SparseCUDA:
+      return Backend::SparseCPU;
+    case Backend::Undefined:
+      return Backend::Undefined;
+    default:
+      AT_ERROR("Unknown backend");
+  }
+}
+
+static inline Backend backendToCUDA(Backend b) {
+  switch (b) {
+    case Backend::CPU:
+      return Backend::CUDA;
+    case Backend::CUDA:
+      return Backend::CUDA;
+    case Backend::SparseCPU:
+      return Backend::SparseCUDA;
+    case Backend::SparseCUDA:
+      return Backend::SparseCUDA;
+    case Backend::Undefined:
+      return Backend::Undefined;
+    default:
+      AT_ERROR("Unknown backend");
+  }
+}
+
+constexpr DeviceType kCPU = DeviceType::CPU;
+constexpr DeviceType kCUDA = DeviceType::CUDA;
+
+static inline const char* toString(Backend b) {
```

Comment:
Should we rename this function? (Maybe not this PR?)

Reply:
what would you like it renamed to?  It kind of nice as is in that there are functions named this for each ATen construct so you can just write at::toString(my_at_thing) and it generally works.

Reply:
I think I generally assume that `operator<<` works for arbitrary types, but not necessarily `toString`. But it doesn't matter too much.